### PR TITLE
Add tidb source code introduction

### PR DESCRIPTION
**Pull request by shenli**
Add a brief document for TiDB's source code.
@ngaut PTAL

### CODE REVIEW

**Comment by shenli** [comment]
@coocood PTAL

File: tidb/sourcecode.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
++ docs
+
+    一些简略的文档，更详细的文档参见[中文文档](https://github.com/pingcap/docs-cn)以及[英文文档](https://github.com/pingcap/docs)。 
+
++ executor
+
+    TiDB 执行器，SQL 语句最终会转化为一系列执行器（物理算子）的组合。这个包对外暴露的最主要的接口是 Executor:
+    ```
+    type Executor interface {
```
Comment (ngaut) [inline tidb/sourcecode.md:45]: 代码格式重新排版一下

Comment (shenli) [reply tidb/sourcecode.md:45]: Fixed. PTAL

**Comment by shenli** [comment]
@ngaut PTAL

**Comment by ngaut** [comment]
Please fix those pictures.

**Comment by shenli** [comment]
@ngaut PTAL

File: tidb/sourcecode.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+整个流程的入口在 tidb 包的 session.go 中，TiDB-Server 调用 Session.Execute() 接口，输入为文本格式的 SQL 语句，实现在 session.Execute()。
+
+首先会调用 Compile() 对 SQL 语句进行语法解析（tidb.Parse()），解析后会得到一个 stmt 的列表，这里每条语句都是一个 AST（抽象语法树），每一个语法单元都是数的一个 Node，结构定义都在 ast 包中。
+
+得到 AST 之后，调用 executor 包中的 Compiler，输入 AST，得到执行器（Compiler.Compile()）。在这个过程中，会完成合法性验证、制定计划以及优化查询计划。
+
+进入 Compiler.Compile() 函数，首先会调用 plan.Validate() 对语句进行合法性验证（见 plan/validator.go），然后进入 Preprocess 流程，目前这个阶段，Preprocess 只做了名称解析工作，及将 SQL 语句中的提到的 column 名或者  alias name 绑定到对应的 field上。比如”select c from t;”这个语句，会将 c 这个名字绑定到 t 这个表的对应列上（具体的实现见 plan/resolver.go）。之后就进入 optimizer.Optimize()。
+
+Optimize() 方法中，首先对 AST 中各个node的结果进行推导，如 "select 1, ’xx‘, c from t;”，对于 select fields，第一个 field是”1“，其类型为 Longlong，第二个 field 是”’xx‘“, 其类型为 VarString，第三个field是 “c” 其类型为表 t 中 column c 的类型。注意这里除了类别之外还有 charset 等信息，都要进行推导，具体实现见 plan/typeinferer.go。完成类型推导后，进行逻辑优化（planBuilderbuild()），主要工作是根据代数运算，对 AST 进行等价变换，化简 AST。比如 ”select c from t where c > 1+1*2;“ 可以等价变换为”select c from t where c > 3;“。
```
Comment (coocood) [inline tidb/sourcecode.md]: planBuilder.build()

File: tidb/sourcecode.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
++ 调用 aggPushDownSolver.aggPushDown，将聚合算子下推到 Join 之前
+
+拿到逻辑优化后的查询计划树之后，会进行物理优化，代码的入口是对逻辑查询计划树的根节点调用 convert2PhysicalPlan(&requiredProperty{})，其中 requiredProperty 是对下层返回结果顺序、行数的要求。
+逻辑查询计划树从根节点开始，不断的递归调用，将每个节点从逻辑算子转成物理算子，并且根据每个节点的查询代价找到一条比较好的查询路径。
+
+# 执行器
+
+虽然优化器是最核心的组件，但是缺少优秀的执行器，依旧无法构成一个优秀的数据库。同样以军队为例，执行器就是军队中冲锋陷阵的士兵。再厉害的将军，如果没有一群能征善战的士兵，同样无法打胜仗。
+
+相比 MySQL，TiDB 的执行器有两个有点，第一是整个计算框架是一个 MPP 的框架，计算会在多台 TiKV 以及 TiDB 节点上进行，尽可能提高效率和速度；第二是单个算子会尽可能并行，比如 Join/Union 等算子，会启动多个线程同时计算，整个数据计算流程构成一个 pipeline，尽可能缩短每个算子的等待时间。所以 TiDB 在处理大量数据时，比 MySQL 表现好。
```
Comment (coocood) [inline tidb/sourcecode.md]: 有点->优点

**Comment by ngaut** [comment]
LGTM

**Pull request merged**